massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1057 @@
+    pub start: MassaTime,
+    /// a timestamp at which the deployment is considered failed (timeout > start)
+    pub timeout: MassaTime,
+}
+
+// Need Ord / PartialOrd so it is properly sorted in BTreeMap
+
+impl Ord for MipInfo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.start, &self.timeout).cmp(&(other.start, &other.timeout))
+    }
+}
+
+impl PartialOrd for MipInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for MipInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+            && self.version == other.version
+            && self.component == other.component
+            && self.start == other.start
+            && self.timeout == other.timeout
+    }
+}
+
+impl Eq for MipInfo {}
+
+// Need to impl this manually otherwise clippy is angry :-P
+impl Hash for MipInfo {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.component.hash(state);
+        self.start.hash(state);
+        self.timeout.hash(state);
+    }
+}
+
+machine!(
+    /// State machine for a Versioning component that tracks the deployment state
+    #[allow(missing_docs)]
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    enum MipState {
+        /// Initial state
+        Defined,
+        /// Past start
+        Started { threshold: Amount },
+        /// Wait for some time before going to active (to let user the time to upgrade)
+        LockedIn,
+        /// After LockedIn, deployment is considered successful
+        Active,
+        /// Past the timeout, if LockedIn is not reach
+        Failed,
+    }
+);
+
+impl Default for MipState {
+    fn default() -> Self {
+        Self::Defined(Defined {})
+    }
+}
+
+#[allow(missing_docs)]
+#[derive(IntoPrimitive, Debug, Clone, Eq, PartialEq, TryFromPrimitive, PartialOrd, Ord)]
+#[repr(u32)]
+pub enum MipStateTypeId {
+    Error = 0,
+    Defined = 1,
+    Started = 2,
+    LockedIn = 3,
+    Active = 4,
+    Failed = 5,
+}
+
+impl From<&MipState> for MipStateTypeId {
+    fn from(value: &MipState) -> Self {
+        match value {
+            MipState::Error => MipStateTypeId::Error,
+            MipState::Defined(_) => MipStateTypeId::Defined,
+            MipState::Started(_) => MipStateTypeId::Started,
+            MipState::LockedIn(_) => MipStateTypeId::LockedIn,
+            MipState::Active(_) => MipStateTypeId::Active,
+            MipState::Failed(_) => MipStateTypeId::Failed,
+        }
+    }
+}
+
+/// A message to update the `MipState`
+#[derive(Clone, Debug)]
+pub struct Advance {
+    /// from VersioningInfo.start
+    pub start_timestamp: MassaTime,
+    /// from VersioningInfo.timeout
+    pub timeout: MassaTime,
+    /// % of past blocks with this version
+    pub threshold: Amount,
+    /// Current time (timestamp)
+    pub now: MassaTime,
+}
+
+impl Default for Advance {
+    fn default() -> Self {
```

Comment:
Given the content of the impl, I think this should only be useful for tests, maybe remove this?

Reply:
This is used in MipStateHistory::new. As it allow use to store an advance in history for state Defined

Reply:
I've removed the Default impl as it might confuse other devs